Replace Ace Editor with Monaco Editor

bennothommo opened the pull request:
Fixes #431. Related: https://github.com/wintercms/wn-builder-plugin/pull/73
<img width="1172" height="722" alt="Screenshot 2026-01-02 at 1 02 11 AM" src="https://github.com/user-attachments/assets/e84089e6-75b5-4621-857e-18b0bd48f324" />
<img width="1326" height="260" alt="Screenshot 2026-01-02 at 1 02 40 AM" src="https://github.com/user-attachments/assets/561a9252-360f-4eed-8532-d1be2013b8cb" />
<img width="1440" height="594" alt="Screenshot 2026-01-02 at 1 03 25 AM" src="https://github.com/user-attachments/assets/5879f3f8-44a5-410e-8e18-e862321f24ea" />

This PR introduces the Monaco editor in replacement of the Ace Editor. At this stage for the PR, we are aiming for parity with the functionality previously provided by Ace, with the intention of increasing the functionality after this PR is merged.

### Still to do:

- [x] Support all prior configuration options.
- [x] Re-implement all previous themes, or find suitable alternatives (see https://github.com/brijeshb42/monaco-themes)
- [x] Ensure code editor works in all current core locations (settings areas, CMS editor, Builder plugin)
- [x] Try and figure out how to remove the automated loading of the Codicon font, as it's being implemented in the widget's CSS assets.

### Changes:
- Auto-completing tags preference is no longer available, as Monaco doesn't have this as an option.
- Autocompletion and code snippets preferences are no longer available, as these are more contextual and will be implemented based on what is being edited in the editor.
- Code folding is now a simple boolean.
- The **Crimson Editor** theme has been dropped - I couldn't find a complete replica of this theme, and it looks too similar to other themes in the list.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Replaced legacy editor with a modern Monaco-based editor: minimap, bracket colors, inline color previews, richer language services, and many new themes.

* **Improvements**
  * Unified editor preferences into a single backend plugin and refreshed editor styling, toolbar, and tab collapse icon behavior.
  * Asset/loading initialization updated for improved editor packaging.

* **Bug Fixes**
  * Fixed duplicated render invocations causing double render handling.

* **Documentation**
  * Added a README for the new editor integration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

bennothommo committed:
Initial setup of Monaco build process

bennothommo committed:
Delete workshop theme

bennothommo committed:
Initial compilation and setup of Monaco

bennothommo committed:
Further work on Monaco Editor support

- Add most configuration options
- Dispose and create editor depending on whether in view
- Fix sizing issue

bennothommo committed:
Ignore Monaco dev module chunks

bennothommo committed:
Add configuration for indent guides and bracket colors

bennothommo committed:
WIP preferences page update

bennothommo committed:
WIP preferences - add code minimap config, fix preview

LukeTowers commented on modules/backend/ServiceProvider.php:
```diff
@@ -104,6 +105,10 @@ protected function registerAssetBundles()
         });
+
+        MixAssets::instance()->registerCallback(function ($mix) {
+            $mix->registerPackage('module-backend.codeeditor', '~/modules/backend/formwidgets/codeeditor/assets/winter.mix.js');
```
```suggestion
            $mix->registerPackage('formwidget.codeeditor', '~/modules/backend/formwidgets/codeeditor/assets/winter.mix.js');
```

bennothommo replied:
The reason I named it that way is that I see the code editor as a "subset" of the Backend module. I was thinking of introducing a way to namespace packages (like this) and add an argument to `mix:compile` to compile everything within a namespace.

LukeTowers replied:
Sure, that sounds fine. Could we do `module.backend.formwidgets.codeeditor` perhaps?

bennothommo replied:
It'd be `module-backend.formwidgets.codeeditor`, but yep, that works for me.

LukeTowers commented on modules/backend/assets/ui/js/ui/WidgetHandler.js:
```diff
@@ -0,0 +1,70 @@
+        this.initialiseWidgets(element);
+    }
+
+    initialiseWidgets(element) {
```
We should stick to the US spelling of initialize: https://github.com/search?q=org%3Awintercms+initialize&type=code vs https://github.com/search?q=org%3Awintercms+initialise&type=code

bennothommo replied:
True - I'll fix the references in the other repos when I can, but I've completed it for this PR.

LukeTowers commented on modules/backend/controllers/preferences/_field_editor_preview.php:
```diff
@@ -1,22 +1,25 @@
       data-margin="0"
-      data-vendor-path="<?= Url::asset('/modules/backend/formwidgets/codeeditor/assets/vendor/ace') ?>/">
-      <textarea name="editorpreferences_codeeditor"><?= e($this->makePartial('example_code')) ?></textarea>
+      data-vendor-path="<?= Url::asset('/modules/backend/formwidgets/codeeditor/assets/vendor/ace') ?>/"
```
Is this line still required?

bennothommo committed:
Add preferences page UI js to build process

bennothommo committed:
Code cleanup, commenting and changing to US spelling

bennothommo committed:
Add word wrap, font size and tab size preferences

bennothommo committed:
Add rest of preferences, add bracket color and color picker preferences

bennothommo committed:
Drop unneeded line

bennothommo committed:
WIP themes

bennothommo committed:
WIP themes

bennothommo committed:
Add ability to preview multiple languages in code editor settings, add theme switch

bennothommo committed:
Add all available themes (only cut Crimson Editor)

bennothommo committed:
Merge branch 'wip/monaco-editor' of github.com:wintercms/winter into wip/monaco-editor

bennothommo committed:
Increase size of code editor preview

bennothommo committed:
Add listener for jQuery render to render widgets

bennothommo committed:
Update Mix alias for code editor

bennothommo committed:
Add local event handler for widgets

bennothommo committed:
Convert editor preview on Preference page to actual widget, add local events for editor

bennothommo committed:
Fix "once" local events

bennothommo committed:
Optimise editor on resize and leaving/entering tab, add fullscreen

bennothommo committed:
Allow ad-hoc Snowboard listeners to function even if no plugins are listening to event

bennothommo committed:
Fix code editor in CMS area, add ability for hidden code areas

bennothommo committed:
Add more reliable method of retrieving widgets based on a given element

bennothommo committed:
Add content manipulation methods in editor

bennothommo committed:
Fix component placement and PHP highlighting in CMS pages

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Allow code editor to work without status bar

bennothommo committed:
Add wrap capability

bennothommo committed:
WIP Markdown Editor rewrite

bennothommo committed:
Recompile assets

bennothommo committed:
Return selections after editor modification, small tweaks

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/monaco-editor

bennothommo committed:
WIP input tracking in Markdown Editor

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Remove call to Theme::getActiveTheme()

This method requires a DB call, which was causing a race condition in certain circumstances as the DB is not initialised at this point. Since we're already scanning all themes for Mix files, we might as well load the Mix assets manually defined at the same time.

bennothommo committed:
Adding formatting wrap/unwrap. Add preview.

bennothommo committed:
WIP add Italic format, improve configuration

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Allow Stripe Loader to be disabled by Snowboard requests

bennothommo committed:
WIP config options

bennothommo committed:
WIP keybindings

bennothommo committed:
Fix multiple backspace issue

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
After building Monaco bundle, remove inline Codicon font calls

bennothommo committed:
Update bundle, fix Italic detection in Markdown editor

bennothommo committed:
Use in-bult VS theme for Markdown editor

It appears to be a lot better for showing Markdown formatting.

bennothommo committed:
Only allow Markdown editor actions to fire if enabled

bennothommo committed:
Allow selection of in-built Monaco themes

bennothommo committed:
Fix resize issues with code editor

- Prevent context error when editor refreshes after resize event
- Fix issue where editor continually increases height when code editor is in a fluid height container (ie. "stretch" fields)

bennothommo committed:
Allow editor to start from a certain line

This allows prepending of content for the code editor that cannot be edited by the end-user.

bennothommo committed:
Round to nearest pixel

LukeTowers commented:
@bennothommo what's the status of this?

bennothommo commented:
When I last worked on this, I was having some dramas with the Markdown editor, since that also had to be converted to Monaco. I was hesitant to include another editor just for the Markdown editor, but if we can find a lean one, I'm happy to use that instead and keep Monaco for just the code editing (which it works amazingly with).

LukeTowers commented:
What's the problem with the markdown editor? Does it currently use ace?

bennothommo commented:
Yeah it uses Ace currently. I feel keeping Ace around and introducing Monaco for the code editor might be too much "bloat".

LukeTowers commented:
I'm fine with having both if we can clearly identify all the locations we can remove Ace from as soon as we get a replacement for the usage of Ace in the markdown editor; that way at least it won't hold us up from getting the benefits of monaco and it'll be simple enough to remove Ace when we replace the markdown editor.

Any thoughts on that?

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Prevent misclicking of outside elements when selecting code inside editor and stopping outside editor

bennothommo committed:
Fix reselection of text after changing bold status

bennothommo committed:
WIP fixing reselection of text after bold/italitc toggle

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Code quality fixes, discard repeated package registrations

In some cases, such as unit tests, a package might be registered twice with the same package name and config. Rather than erroring out, we'll just discard the repeated registration. Note that if a config file is registered twice with different package names, or two config files are registered to the same package name, this is still an error.

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Merge branch 'develop' into wip/monaco-editor

bennothommo committed:
Add ability to modify CMS component templates inside Monaco

bennothommo committed:
Fix exception on AJAX requests made by widgets when template form is empty

As the code editor fires an AJAX call immediately on load to get the theme, the CMS module has to bind the form widget to the controller. Originally, this simply loaded the template, but because the file name of the template likely is blank, this would fail the validation on the file name. To resolve this, we'll simply "create" the template if no template path exists.

bennothommo committed:
Remove old token expansion functionality

LukeTowers commented on modules/backend/formwidgets/codeeditor/assets/js/codeeditor.old.js:
Is this still required?

bennothommo replied:
I'll need it probably up until the last moment, just in case there's any other original functionality I need to port across. I'll remove it once we're ready to go.

LukeTowers commented on modules/backend/formwidgets/codeeditor/assets/less/codeeditor.less:
```diff
@@ -1,103 +1,665 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
```
Can we import this from a separate file?

bennothommo replied:
I can have another look, but from what I remember, the CSS is inlined in Monaco so it may not be importable.

LukeTowers commented on modules/backend/models/Preference.php:
```diff
@@ -59,6 +59,8 @@ public function initSettingsData()
         $this->editor_font_size = $config->get('editor.font_size', 12);
         $this->editor_word_wrap = $config->get('editor.word_wrap', 'fluid');
         $this->editor_code_folding = $config->get('editor.code_folding', 'manual');
+        // @deprecated v1.2.2
```
```suggestion
        // @deprecated v1.3.0
```

LukeTowers commented on modules/system/assets/js/snowboard/main/Snowboard.js:
```diff
@@ -336,7 +336,6 @@ export default class Snowboard {
         const listeners = this.listensToEvent(eventName);
         if (listeners.length === 0) {
             this.debug(`No listeners found for global event "${eventName}"`);
-            return true;
```
Is this intentional?

bennothommo replied:
Yes, it short-circuits the event listeners if there are none available.

LukeTowers commented on modules/system/assets/js/snowboard/main/Snowboard.js:
```diff
@@ -437,7 +436,6 @@ export default class Snowboard {
         const listeners = this.listensToEvent(eventName);
         if (listeners.length === 0) {
             this.debug(`No listeners found for global promise event "${eventName}"`);
-            return Promise.resolve();
```
Is this intentional?

bennothommo replied:
Yes, it short-circuits the event listeners if there are none available.

LukeTowers commented on modules/system/assets/ui/font/fa-brands-400.ttf:
Why are these files being added?

bennothommo replied:
Possibly done before we changed how Font Awesome was being built. I'll remove.

LukeTowers commented on themes/demo/pages/ajax.htm:
Unintentional changes to this file?

bennothommo replied:
Most likely - I'll revert them.

bennothommo committed:
Update modules/backend/models/Preference.php

Co-authored-by: Luke Towers <git@luketowers.ca>

LukeTowers commented:
@coderabbitai review

LukeTowers committed:
Merge branch 'develop' into wip/monaco-editor

# Conflicts:
#	modules/backend/formwidgets/codeeditor/assets/css/codeeditor.css
#	modules/system/assets/ui/storm.css

LukeTowers committed:
Fix issue with closing full screen mode on page with markdown editor as well as codeeditor

LukeTowers committed:
Add support for JSON themes

Also improved the UX for previewing the code editor settings, added one-dark-pro and winter themes, renamed codeeditor's source JS file to codeeditor.js, fixed a bug with alt key binding, and removed unused code from the Preferences controller.

LukeTowers committed:
Darken the one dark pro theme background for readability

LukeTowers committed:
Use asset URL for codeeditor assets

LukeTowers committed:
Improve removal of inline codicon font-face

LukeTowers committed:
Restore original ace-backed codeeditor for RichEditor to use

RichEditor is currently using Froala v2.9.3 which is difficult to customize as we don't have the original DRM files anymore. Ideally we would like to replace the codeeditor view of Froala with Monaco, but for now we'll retain ace for this use case.

LukeTowers committed:
Cleanup

LukeTowers committed:
Deconflict ace-codeeditor from monaco codeeditor

LukeTowers committed:
Discard changes to modules/backend/formwidgets/markdowneditor/assets/less/markdowneditor.less

LukeTowers committed:
Discard changes to modules/backend/formwidgets/markdowneditor/partials/_markdowneditor.php

LukeTowers committed:
Discard changes to themes/demo/pages/ajax.htm

LukeTowers committed:
Restore Ace Editor for MarkdownEditor

Markdown editor needs some more work, we might look at using TipTap or Milkdown / Crepe in the future.

LukeTowers committed:
Move asset bundle registration to the boot() method

Interacting with the PackageManager triggers a need to use the database connection to get the active theme, the database connection is not available until the boot process.

LukeTowers committed:
Fix styling issue with tab collapse icon on fancy layouts

LukeTowers committed:
Improve codeeditor bundle loading

LukeTowers committed:
Tidying

LukeTowers committed:
Add asset_url link tag to backend layouts to make Snowboard.url().asset() function

LukeTowers committed:
Improve monaco asset loading to support cross origin asset URLs and reduce issues caused by timing issues with webpack loading

LukeTowers committed:
Improve backwards compatibility for the codeeditor

LukeTowers committed:
Add setMarkers and setDecorations helper to the CodeEditor Snowboard plugin

LukeTowers committed:
Compile assets

LukeTowers committed:
ONE FREAKING COMMA AND THE WORLD BLOWS UP

For anyone else trying to track down self being clobbered, this is very handy:

(function () {
  try {
    const desc = Object.getOwnPropertyDescriptor(window, 'self');
    console.log('window.self descriptor:', desc);

    let _selfValue = window;
    Object.defineProperty(window, 'self', {
      configurable: true,
      get() { return _selfValue; },
      set(v) {
        console.error('window.self set to:', v);
        console.trace('stack:');     // <- prints the exact script/line
        _selfValue = v;
      },
    });
  } catch (e) {
    console.warn('Failed to trap window.self:', e);
  }
})();

LukeTowers committed:
Recompile assets

LukeTowers committed:
Add docs for monaco powered codeeditor

LukeTowers committed:
Move theme loading to directly in JS instead of over AJAX

Also added build cleanup step to build process to avoid leaving behind artifacts from previous builds.

LukeTowers committed:
Update docs for the codeeditor

LukeTowers committed:
Merge branch 'develop' into wip/monaco-editor

# Conflicts:
#	modules/backend/assets/vendor/ace-codeeditor/build-min.js

LukeTowers committed:
Merge branch 'develop' into wip/monaco-editor

LukeTowers committed:
Improve type hinting of CMS Index controller

LukeTowers committed:
Fix incorrect codeeditor language being set for .txt & .js files in the CMS

LukeTowers committed:
Remove console.log statements

LukeTowers committed:
Remove broken updateFormEditorMode

Code didn't really do anything even with the aceeditor.

LukeTowers committed:
Coderabbit picky

LukeTowers committed:
Add TODO for JS that might interact with ace

LukeTowers committed:
Add Monaco compatibility for the drag.value.js file

In theory only used by the Winter.Notify plugin, and even then, only if someone has a custom action defined that makes use of a codeeditor field.

LukeTowers committed:
Improve CMS page handling of code sections

LukeTowers commented on modules/backend/formwidgets/codeeditor/assets/js/codeeditor.js:
```diff
@@ -1,468 +1,1531 @@
+         * @param {(model: monaco.editor.ITextModel, token: monaco.CancellationToken) => Promise<monaco.languages.CodeLens[]>} provider
+         * @param {(model: monaco.editor.ITextModel, codeLens: monaco.languages.CodeLens, token: monaco.CancellationToken) => Promise<monaco.languages.CodeLens>} resolver
+         */
+        addCodeLens(language, provider, resolver = null) {
+            this.disposables.push(
+                monaco.languages.registerCodeLensProvider(language, {
+                    provideCodeLenses: (model, token) => {
+                        return provider(model, token);
+                    },
+                    resolveCodeLens: (model, codeLens, token) => {
+                        return resolver(model, codeLens, token) ?? codeLens;
+                    },
+                })
+            );
+        }
```
Callers need to implement the command. Not fixing this for now.
